R/readChromData.R

```diff
@@ -0,0 +1,246 @@
+                    stop("Got more redundant chromatograms than expected")
+                res_chrs[[idx_to_place]] <- Chromatogram(
+                    rtime = chr_data[[i]][, 1],
+                    intensity = chr_data[[i]][, 2],
+                    precursorMz = hdr$precursorIsolationWindowTargetMZ[i],
+                    productMz = hdr$productIsolationWindowTargetMZ[i],
+                    fromFile = idx)                
+            }
+            res_chrs
+        }), use.names = FALSE)
+
+    Chromatograms(chrs, phenoData = pdata, featureData = fdata,
+                  ncol = length(files))
+}
+
+#' @title Combine `data.frame`s without collapsing non-unique rows
+#'
+#' @description
+#'
+#' Combine a list of `data.frame`s into a single `data.frame`. In
+#' contrast to a *unique* combination in which each unique row will only be
+#' present once, this function does not reduce redundant rows to a single row,
+#' but returns them. In other words, if three rows in one of the input
+#' `data.frame` contain the same values, the result `data.frame` also has the
+#' same 3 rows, while `unique` would collapse them into a single row. If rows
+#' with the same values are present multiple times in multiple input
+#' `data.frame`'s, e.g. 3 times in one `data.frame` and 2 times in
+#' another one, the result `data.frame` will contain the rows the maximum time
+#' they are present across tables. See examples for details.
+#'
+#' @note
+#'
+#' The ordering of a combined `data.frame` from `data.frame`s with some or
+#' all columns being `factors` might not be as expected. It is thus suggested
+#' to avoid passing `data.frame`s with `factors` to the function.
+#' 
+#' @param x
+#'
+#' @param cols `character` defining the columns that should be returned.
+#'
+#' @return `data.frame`
+#' 
+#' @noRd
+#'
+#' @md
+#' 
+#' @author Johannes Rainer
+#'
+#' @examples
+#'
+#' A <- data.frame(a = c(1, 1, 2, 3, 4, 5, 6), b = c(2, 2, 3, 4, 5, 5, 6))
+#' B <- data.frame(a = c(1, 3, 3, 3, 3, 4, 5), b = c(2, 4, 4, 4, 5, 5, 5))
+#' C <- data.frame(a = c(3, 3, 4, 4), b = c(4, 5, 5, 5))
+#'
+#' x <- list(A, B, C)
+#'
+#' .combine_data.frame(x)
+#'
+#' D <- data.frame(a = c("z", "b", "a", "g", "g"),
+#'                 b = c(1, 2, 3, 4, 4), stringsAsFactors = FALSE)
+#' E <- data.frame(a = c("g", "a", "d", "g"),
+#'                 b = c(4, 3, 1, 1), stringsAsFactors = FALSE)
+#' .combine_data.frame(list(D, E))
+.combine_data.frame <- function(x, cols) {
+    if (!length(x))
+        stop("length of 'x' must be > 0")
+    if (!all(unlist(lapply(x, is.data.frame))))
+        stop("all elements in 'x' need to be a data.frame")
+    fd <- do.call(rbind, x)
+    if (missing(cols))
+        cols <- colnames(fd)
+    else {
+        if (!all(cols %in% colnames(fd)))
+            stop("All columns specified with 'cols' have to be present in the",
+                 " data.frames")
+    }
+    ## Make first unique elements
+    fd <- unique(fd[, cols, drop = FALSE])
```

Comment:
Thanks for the suggestion - I'll check it (also in terms of performance).

Reply:
Ah, you didn't like the snake_case variable names in the original code, that's why you want to change it ;)

Reply:
yes, your's is also hard to understand - I failed to make it running on the example data set - could you please push a `.combine_data.frame2` function so I could test?

Reply:
> could you please push a .combine_data.frame2 function so I could test?

Done.

> yes, your's is also hard to understand

In fact I just create the `data.frame` id to avoid one loop. Could be slightly faster. It works at least for your `list(A, B, C)` and `list(D, E)` test cases. BTW in the unit test you create `list(A, B, C)` and `exp` but never compare the results.

> Ah, you didn't like the snake_case variable names in the original code, that's why you want to change it ;)

You got me :wink: